packages/compiler/core/parser.ts

```diff
@@ -945,6 +955,37 @@ export function parse(code: string | SourceFile, options: ParseOptions = {}): Ca
     };
   }
 
+  function parseAugmentDecorator(): AugmentDecoratorStatementNode {
+    const pos = tokenPos();
+    parseExpected(Token.AtAt);
+
+    // Error recovery: false arg here means don't treat a keyword as an
+    // identifier. We want to parse `@ model Foo` as invalid decorator
+    // `@<missing identifier>` applied to `model Foo`, and not as `@model`
+    // applied to invalid statement `Foo`.
+    const target = parseIdentifierOrMemberExpression(undefined, false);
+    const args = parseOptionalList(ListKind.DecoratorArguments, parseExpression);
+    if (args.length === 0) {
+      error({ code: "augment-decorator-target" });
+    }
+    let [targetEntity, ...decoratorArgs] = args;
+    if (targetEntity.kind !== SyntaxKind.TypeReference) {
+      error({ code: "augment-decorator-target", target: targetEntity });
+      targetEntity = {
+        kind: SyntaxKind.TypeReference,
+        target: createMissingIdentifier(),
+        arguments: [],
```

Comment:
I don't think we should drop the other args in this case. This feels like maybe we will get better experience if this is handled as a checking rule than a parsing rule.

Reply:
Maybe it would be fine to simply take the arguments including the first-non-type-reference as the arguments?

Reply:
In general it's good if the parse tree preserves as much data as possible in error cases. We've had some bugs where I had to go back and fix that up to resolve.

Reply:
yeah my bad here, meant to do that.